`scheduleUsed_corres`

michaelmcinerney opened the pull request:
This proves `scheduleUsed_corres`.

I've rewritten `schedule_used` in the abstract to be monadic. I had hoped that the changes to AInvs would be minimal, but it turned out to be quite a lot of reorganising. Most of the more difficult word proofs (`schedule_used_no_overflow`, `schedule_used_odered_disjoint`, `schedule_used_window`) are pretty much unchanged. Some of the helper lemmas (`refill_budget_check_ordered_disjoint_helper'`, etc.) are new and replace older, clunkier versions. Nothing very exciting is going on in AInvs.

As expected after rewriting `schedule_used`, the `corres` rule was very straightforward. I had to slightly modify `refillAddTail_corres` to allow it to be applied in my lemma. This did not break anything in this branch, but perhaps it is being used elsewhere

corlewis commented on proof/invariant-abstract/DetSchedSchedule_AI.thy:
```diff
@@ -8244,20 +8258,21 @@ lemma refill_update_valid_refills:
-  apply (cases refills)
-   apply (clarsimp simp: schedule_used_def)
+lemma schedule_used_sum:
+  "\<lbrace>\<lambda>s. pred_map (\<lambda>cfg. refills_sum ((scrc_refills cfg) @ [new]) = t) (sc_refill_cfgs_of s) (cur_sc s)
+        \<and> sc_ptr = cur_sc s\<rbrace>
+   schedule_used sc_ptr new
+   \<lbrace>\<lambda>_ s. pred_map (\<lambda>cfg. refills_sum (scrc_refills cfg) = t) (sc_refill_cfgs_of s) (cur_sc s)\<rbrace>"
```
It's very hard to say without actually playing around with it, but would it be easier if these lemmas about `schedule_used` were generalised away from `cur_sc`?

michaelmcinerney replied:
It certainly wouldn't be any easier if these weren't `cur_sc`. I found it very annoying to get this all right, and I dreaded trying to use more general versions along with `wps` (which I've never had a good handle on). If you'd like me to take another look, I can, or maybe I can make a JIRA issue for it

corlewis commented on proof/invariant-abstract/DetSchedSchedule_AI.thy:
```diff
@@ -8329,153 +8344,82 @@ lemma valid_refills_r_amount_bounded_max_sc_period:
-  apply (prop_tac "sum_list (map unat (map r_amount [last lista])) = unat (r_amount (last lista))")
-    apply simp
-  apply (metis sum_list_append map_append)
+lemma unat_max_time:
+  "2 ^ LENGTH(64) - 1 = unat max_time"
```
This and the next lemma should probably be about `max_word` instead of `max_time`, and also marked for being moved (possible even into `Word_Lib`, although it's hard to tell whether they're distinct enough from the already existing lemmas)

michaelmcinerney replied:
Sure. I've got some nice rules generalised to `'a :: len word`. Maybe `Word_Lemmas_Internal.thy` in `Word_Lib`?

lsf37 replied:
Yes, the `*_Internal` theories are the right place to put these.

lsf37 replied:
Apart from this one, of course :-) (sorry, saw that as soon as I wrote it). `max_time` is not going to be defined there yet, so somewhere close to where `max_time` is defined would be a good spot.

lsf37 replied:
(when generalised to `max_word`)

corlewis reviewed (approved):
It's a bit sad to get rid of the nice abstract version of `schedule_used`, but this definitely seems like the best way forward as opposed to doing something more complicated in Refine. Great work!

If possible I always like to properly generalise lemmas for all objects instead of just the `cur_sc` (/`cur_tcb`), but that might not be necessary for this PR.

michaelmcinerney commented on proof/refine/ARM/SchedContextInv_R.thy:
```diff
@@ -1012,6 +1012,120 @@ lemma refillUpdate_corres:
+  done
+
+(* FIXME RT: move? *)
+lemma corres_if4:
+  "\<lbrakk>G = G'; G \<Longrightarrow> corres_underlying sr nf nf' r P P' a c; \<not>G \<Longrightarrow> corres_underlying sr nf nf' r P P' b d\<rbrakk>
+   \<Longrightarrow> corres_underlying sr nf nf' r P P' (if G then a else b) (if G' then c else d)"
```
I suppose this isn't needed. I could just use `corres_if3` (maybe with some sort of `where blah, simplified blah` statement). But this rule is exactly what my (probably eccentric) thought process wanted. Should I get rid of it? Or move it to `Corres_UL.thy`?

corlewis replied:
Eh, if it's more convenient then I'm fine with keeping this one, with a preference for moving it to `Corres_UL`. It just feels a bit strange with how close it is to `corres_if3`.

michaelmcinerney commented:
I've updated and rebased this PR.

Git pretty badly mangled the rebase. It ended up deleting quite a few lemmas from DetSchedScheulde, but I've added those in a separate commit. And considering there were a number of things I wanted to modify in Refine, and the rebase there was not done well automatically, I decided to remove the old commit for Refine and make a new one. Apologies if this makes reviewing more difficult. I've cherry-picked a commit for several word lemmas from the `us_to_ticks` PR, and put `corres_if4` into its own commit, too

mktnk3 commented on lib/Corres_UL.thy:
```diff
@@ -537,6 +537,13 @@ lemmas corres_when2 =
                 and Q="\<top>" and Q'="\<top>" and r=dc, unfolded if_apply_def2, simplified,
                 folded when_def]
 
+lemma corres_if4:
+  "\<lbrakk> G = G';
+     G \<Longrightarrow> corres_underlying sr nf nf' r P P' a c;
+     \<not>G \<Longrightarrow> corres_underlying sr nf nf' r P P' b d \<rbrakk>
+   \<Longrightarrow> corres_underlying sr nf nf' r P P' (if G then a else b) (if G' then c else d)"
+  by (clarsimp simp: corres_underlying_def)
+
```
Hmm. This looks quite trivial to me. `simp` or `clarsimp` will split the conclusion into a conjunction of the two implications in the assumption or something similar. So, `clarsimp` and `conjI/impI` would be pretty much the same thing as applying this rule, woundn’t it?

michaelmcinerney replied:
It is quite trivial. `clarsimp` seems to be greedy and will pull apart all `if`s. `schedule_used` contains three `if`s, so here it will create duplication which I can't see how to handle easily, and disturbs the structure of the proof I was wanting

mktnk3 replied:
Ok, you want to proceed one layer at a time. Did you try using `corres_guard_imp` and `corres_if3`? Or `corres_guard_cong`?

michaelmcinerney replied:
I could use `corres_guard_imp` and `corres_if3`, but I was aiming for a more "fowards" style, where the guards are either provided by strongest postcondition rules, "skip" rules, comparison with the initial guards, or (as a last resort) manual instantiation. `corres_if4` is exactly 

```
corres_if3[where P=T and Q=T for T, where P'=T' and Q'=T' for T', simplified]
```

so I could just use that instead, but I thought that a simple rule like this might be more convenient for people who like this (even more) forwards reasoning style in `corres` proofs

michaelmcinerney replied:
I would certainly prefer to keep it. `corres_if_split` isn't taken. Would you be happy for me to rename it to `corres_if_split`?

mktnk3 replied:
that would be good, if there’s no objection from other people.

mktnk3 commented on proof/refine/ARM/SchedContextInv_R.thy:
```diff
@@ -1268,6 +1268,98 @@ lemma headInsufficientLoop_corres:
+  apply (clarsimp simp: obj_at'_def)
+  done
+
+lemma length_refills_map_is_scRefillCount:
+  "sc_valid_refills' sc
+   \<Longrightarrow> (length (refills_map (scRefillHead sc) (scRefillCount sc) (scRefillMax sc) (scRefills sc)))
+        = scRefillCount sc"
+  apply (clarsimp simp: valid_sched_context'_def refillTailIndex_def refills_map_def)
+  done
```
We already have
```
lemma length_refills_map[simp]:
  "⟦ mx ≤ length list; count ≤ mx ⟧ ⟹ length (refills_map start count mx list) = count"
  by (clarsimp simp: refills_map_def)
```
in the simp set. Given that `sc_valid_refills’` is an abbreviation, I don’t see why we need to have a separate lemma for this particular instance.

michaelmcinerney replied:
Removed

michaelmcinerney commented on proof/invariant-abstract/DetSchedSchedule_AI.thy:
```diff
@@ -8305,153 +8320,82 @@ lemma valid_refills_r_amount_bounded_max_sc_period:
-  apply (prop_tac "sum_list (map unat (map r_amount [last lista])) = unat (r_amount (last lista))")
-    apply simp
-  apply (metis sum_list_append map_append)
+lemma unat_max_time:
+  "2 ^ LENGTH(64) - 1 = unat max_time"
+  apply (clarsimp simp: max_word_def)
   done
```
Yes. Removed

mktnk3 reviewed (approved):
Great to see the monadic version of `schedule_used` working nicely. A few more comments for moving smaller lemmas out of DetSchedSchedule_AI.
